src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,443 @@
+
+        private async ValueTask ComputeViewStateAsync(CancellationToken cancellationToken)
+        {
+            // We do not want this work running on a background thread
+            await TaskScheduler.Default;
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Do the expensive LSP work of actually computing the items to show.
+            var (documentSymbolData, newTextSnapshot) = await ComputeDocumentSymbolDataAsync(cancellationToken).ConfigureAwait(false);
+
+            // Now, go back to the UI and grab the prior view state we set, and the current UI values we want to update the data with.
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var searchText = this.SearchText;
+            var sortOption = this.SortOption;
+            var lastPresentedViewState = this.LastPresentedViewState;
+
+            // Jump back to the BG to do all our work.
+            await TaskScheduler.Default;
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var searchTextChanged = searchText != lastPresentedViewState.SearchText;
+            var oldViewModelItems = lastPresentedViewState.ViewModelItems;
+
+            // if we got new data or the user changed the search text, recompute our items to correspond to this new state.
+            // Apply whatever the current search text is to what the model returned, and produce the new items.
+            var newViewModelItems = GetDocumentSymbolItemViewModels(
+                sortOption, SearchDocumentSymbolData(documentSymbolData, searchText, cancellationToken));
+
+            // If the search text changed, just show everything in expanded form, so the user can see everything
+            // that matched, without anything being hidden.
+            //
+            // in the case of no search text change, attempt to keep the same open/close expansion state from before.
+            if (!searchTextChanged)
+            {
+                ApplyOldStateToNewItems(
+                    oldSnapshot: lastPresentedViewState.TextSnapshot,
+                    newSnapshot: newTextSnapshot,
+                    oldItems: oldViewModelItems,
+                    newItems: newViewModelItems);
+            }
+
+            // Now create an interval tree out of the view models.  This will allow us to easily find the intersecting
+            // view models given any position in the file with any particular text snapshot.
+            var intervalTree = SimpleIntervalTree.Create(new IntervalIntrospector(), Array.Empty<DocumentSymbolDataViewModel>());
+            AddToIntervalTree(newViewModelItems);
+
+            var newViewState = new DocumentOutlineViewState(
+                newTextSnapshot,
+                searchText,
+                newViewModelItems,
+                intervalTree);
+
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
Again, why do we need to switch to the UI thread in the View Model? Only the View should be accessing data that is thread afinitialized.

Reply:
discussed offline.